### PR TITLE
Prevent setting bins unit

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Resolving a small TODO.
- Add `events` property. This is partially (and may fully) replace `array.bins.constituents`, but for now I am keeping both, since there are more things to consider. I suggest we try this in practice for a while.

### CODE REVIEW

**Review by jl-wynen — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
Hmm, this is actually wrong (or the recommendation in the exception is): `array.bins.data` is also binned data (containing not `DataArray` elements but `Variable` elements). Three options:
- Set bin-element unit if `array.bins.data.unit = 'm'` is called.
- Always set bin-element unit if unit is set: `array.unit = 'm'` would set the inner unit. This feels a bit odd, but maybe it is safe?
- Set unit via flat buffer: `array.bins.constituents['data'].unit = 'm'`. This is the safest option. Right now a bit cumbersome, but I am considering adding an `events` property in parallel to the `bins` property: `array.events.unit = 'n'` where `array.events` is what now is `array.bins.constituents['data']`.

**Comment by jl-wynen** [comment]
How would `array.unit = 'm'` behave regarding slicing? If the bins cover only part of the buffer, would this throw or succeed? This kind of check sounds pretty expensive. So in order to be consistent with other places where we forbid unit changes on slices, we should just forbid it for binned arrays altogether (through the outer Variable).

`array.bins.constituents['data'].unit = 'm'` is really verbose but given the above comment, that is maybe not a bad thing.
Having an `events` property would nice and then we can also allow `array.events.unit = 'm' `.

**Comment by SimonHeybrock** [comment]
> How would `array.unit = 'm'` behave regarding slicing? If the bins cover only part of the buffer, would this throw or succeed? This kind of check sounds pretty expensive. So in order to be consistent with other places where we forbid unit changes on slices, we should just forbid it for binned arrays altogether (through the outer Variable).

It would do a (cheap) check on `m_buffer` in the model, which would NOT detect if only certain bins are included. So I think you are right that this may not be a good idea.

> `array.bins.constituents['data'].unit = 'm'` is really verbose but given the above comment, that is maybe not a bad thing.
> Having an `events` property would nice and then we can also allow `array.events.unit = 'm' `.

... this would suffer from the same "problem" as above: affecting also areas of the buffer not referenced by any of the bins. Is this an issue in practice?

**Comment by jl-wynen** [comment]
Is `.bins.constituents['data']` sliced or is it a view on the whole Variable? But regardless, there can be gaps which would not be handled by a slice view. So I would regard `.bins.constituents['data']` as the actual underlying buffer. And then allowing a unit change makes sense.
Or am I getting this wrong?

**Comment by SimonHeybrock** [comment]
> Is `.bins.constituents['data']` sliced or is it a view on the whole Variable?

Could be either, typically not sliced.

> But regardless, there can be gaps which would not be handled by a slice view. So I would regard `.bins.constituents['data']` as the actual underlying buffer. And then allowing a unit change makes sense.
> Or am I getting this wrong?

The problem is that when you slice a binned variable, both the original and the sliced object refer to the full (typically unsliced) buffer, and as you say the only way of checking whether this is the case would be to check all the indices.... but that would maybe not be what we want, e.g., if we want to implement over-allocation which may deliberately leave gaps between bins.

**Comment by jl-wynen** [comment]
A slice of a binned variable in an outer dim would be detectable, right? We could use that to prevent setting the unit.
In all other cases I think it would be ok to allow it. Otherwise there would be no way to change the unit.

**Comment by SimonHeybrock** [comment]
> A slice of a binned variable in an outer dim would be detectable, right? We could use that to prevent setting the unit.
> In all other cases I think it would be ok to allow it. Otherwise there would be no way to change the unit.

Yes, this is detectable. But if we provide direct access to the buffer the user could bypass such a check. But we have the same conceptual loophole if someone modifies elements that are not referenced by the bins in the slice.

**Comment by SimonHeybrock** [comment]
Updated PR text:
>     * Add `events` property. This is partially (and may fully) replace `array.bins.constituents`, but for now I am keeping both, since there are more things to consider. I suggest we try this in practice for a while.

File: python/src/scipp/_bins.py [hunk 1]
```diff
@@ -151,6 +151,20 @@ def concatenate(self, dim):
         return self._obj.concatenate(dim)
 
 
+def _events(obj):
+    """
+    Returns the underlying data the bins stored in the object, or None if the
```
Comment (jl-wynen) [inline python/src/scipp/_bins.py:156]: "Returns the underlying data _of_ the bins stored in the object"

**Review by jl-wynen — CHANGES_REQUESTED** [review]

**Review by jl-wynen — APPROVED** [review]

**Pull request merged**